torch_geometric/nn/summary.py

```diff
@@ -0,0 +1,156 @@
+from collections import defaultdict
+from time import perf_counter
+from typing import Any, Sequence
+
+import torch
+import torch.nn as nn
+
+
+def summary(
+    model: nn.Module,
+    *inputs: Sequence[Any],
+    device: torch.device = 'cpu',
+    max_depth: int = 3,
+) -> str:
+    r"""Summarizes the given PyTorch model. Summarized information includes
+    (1) Layer names, (2) Input/output shapes,
+    (3) # of parameters, and (4) Excutation time.
+
+    .. code-block:: python
+
+        import torch
+        from torch_geometric.nn import summary
+        from torch_geometric.nn.models import GCN
+
+        model = GCN(128, 64, 2, out_channels=32)
+        x = torch.randn(100, 128)
+        edge_index = torch.randint(100, size=(2, 20))
+        print(summary(model, x, edge_index))
+
+    Args:
+        model (nn.Module): PyTorch model to summarize.
+        inputs (Sequence[Any]): Arguments for the model's
+            forward pass.
+        device (torch.device): Device for model and input_data.
+            (default: :obj:`"cpu"`)
+        max_depth (int): Depth of nested layers to display.
+            Nested layers below this depth will not be displayed
+            in the summary. (default: :obj:`"3"`)
+    """
+    def register_pre_hook(info):
+        def hook(module, input):
+            info['input_time'].append(perf_counter())
+
+        return hook
+
+    def register_hook(info):
+        def hook(module, input, output):
+            info['input_shape'].append(get_shape(input))
+            info['output_shape'].append(get_shape(output))
+            info['output_time'].append(perf_counter())
```

Comment:
First save `output_time` and then remaining stats. We don't want to benchmark `get_shape()` here :)